sdk/forms/src/main/java/com/klaviyo/forms/KlaviyoWebViewDelegate.kt

```diff
@@ -15,129 +14,243 @@ import androidx.core.content.ContextCompat.startActivity
+            WebViewCompat.addDocumentStartJavaScript(webView, loadScript, allowedOrigin)
+        }
 
-    @SuppressLint("SetJavaScriptEnabled")
-    private fun buildWebView(): WebView = WebView(Registry.config.applicationContext).also {
-        it.setBackgroundColor(Color.TRANSPARENT)
-        it.webViewClient = this
-        it.settings.userAgentString = DeviceProperties.userAgent
-        it.settings.javaScriptEnabled = true
-        it.settings.domStorageEnabled = true
-        if (isFeatureSupported(WEB_MESSAGE_LISTENER)) {
-            Registry.log.verbose("$WEB_MESSAGE_LISTENER Supported")
-            WebViewCompat.addWebMessageListener(
-                it,
-                IAF_BRIDGE_NAME,
-                setOf(Registry.config.baseCdnUrl),
-                this
-            )
+        if (LOAD_FROM_FILE) {
+            webView.loadUrl(templateUrl.toString())
         } else {
-            Registry.log.verbose("$WEB_MESSAGE_LISTENER Unsupported")
-            it.addJavascriptInterface(this, IAF_BRIDGE_NAME)
+            val html = Registry.config.applicationContext
+                .assets
+                .open("InAppFormsTemplate.html")
+                .bufferedReader()
+                .use(BufferedReader::readText)
+                .replace("SDK_NAME", Registry.config.sdkName)
+                .replace("SDK_VERSION", Registry.config.sdkVersion)
+                .replace("BRIDGE_NAME", BRIDGE_NAME)
+                .replace("BRIDGE_HANDSHAKE", handShakeData)
+                .replace("KLAVIYO_JS_URL", klaviyoJsUrl.toString())
+
+            webView.loadDataWithBaseURL(Registry.config.baseUrl, html, MIME_TYPE, null, null)
+        }
+
+        handshakeTimer?.cancel()
+        handshakeTimer = Registry.clock.schedule(Registry.config.networkTimeout.toLong()) {
+            Registry.log.debug("IAF WebView Aborted: Timeout waiting for Klaviyo.js")
+            close()
         }
     }
 
+    /**
+     * When [DOCUMENT_START_SCRIPT] is unsupported, we must inject our script [onPageStarted] instead
+     */
+    override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) = view
+        ?.takeUnless { !LOAD_FROM_FILE || isFeatureSupported(DOCUMENT_START_SCRIPT) }
+        ?.let { webView ->
+            Registry.log.verbose("$DOCUMENT_START_SCRIPT Unsupported")
+            webView.evaluateJavascript(loadScript) { }
+        } ?: Unit
+
+    /**
+     * Called when loading a resource encounters http status code >= 400
+     */
     override fun onReceivedHttpError(
         view: WebView?,
         request: WebResourceRequest?,
         errorResponse: WebResourceResponse?
     ) {
-        Registry.log.error("HTTP Error: ${errorResponse?.statusCode} - ${request?.url}")
+        Registry.log.warning("HTTP Error: ${errorResponse?.statusCode} - ${request?.url}")
         super.onReceivedHttpError(view, request, errorResponse)
     }
 
-    override fun shouldInterceptRequest(
```

Comment:
I don't mind leaving this in, I deleted it mistakenly in all the font experimentation.